pkg/threescale/backend/README.md

```diff
@@ -0,0 +1,105 @@
+# 3scale Backend Cache
+
+This document briefly explains the functionality offered by the `Backend` construct in this library.
+
+## Usage
+
+`Backend` provides an implementation of the [3scale client interface](https://github.com/3scale/3scale-go-client/blob/v0.4.0/threescale/client.go#L8-L21)
+and so can be used as a drop in replacement for [3scale-go-client](https://github.com/3scale/3scale-go-client)
+anywhere in which out of the box caching is desired between a gateway and Apisonator.
+
+## Data Storage
+
+Currently, the `Backend` supports and ships with an in-memory hash map data storage solution, however it has been left
+open to extension, and is pluggable to custom storage implementations.
+
+## Internals
+
+The cache stores an `Application`, which is a local representation of a model in 3scale itself.
+It is stored under a key which is generated from a combination of the models service identifier and application identifier.
+
+An Application houses the following:
+
+- `LimitCounter` - A counter which tracks rate limited metrics - this maps a metric onto a map of supported time periods,
+which in turn map to a limit (the counter itself) which has a current and max value (integers). It manages a view
+of local state, which incoming requests get recorded to in each iteration. It also holds a view of the last known state in 3scale.
+
+- `UnlimitedCounter` - A counter which tracks metrics with no rate limits attached. It simply maps the metric name to an integer.
+
+An `Application` is capable of holding both read and write locks, allowing it to block where appropriate.
+
+## Functionality
+
+### Supported Endpoints
+
+`Backend` support Authorize, AuthRep and Report functionality. In each case, if the service is unknown to the cache,
+it makes a request to the remote 3scale instance to fetch state. Requests that fail or otherwise error, will not have
+their details added to the cache at that time.
+
+#### Authorization
+
+The following diagram describes the Authorization flow, where the incoming request metrics evaluates against the cache
+to determine if rate limits have been breached, before being authorized or denied.
+
+![Authorization Flow](https://app.lucidchart.com/publicSegments/view/1e274063-8f1e-4adf-b342-1b65efe0de0b/image.png)
+
+#### AuthRep
+
+The following diagram describes the AuthRep flow, where the incoming request metrics evaluates against the cache
+to determine if rate limits have been breached, before being authorized or denied. If authorized, metrics get written
+to the cache.
+
+![AuthRep Flow](https://app.lucidchart.com/publicSegments/view/5e7e18e8-04a0-462d-8e8c-157e4755a355/image.png)
```

Comment:
Minor nit: the output should be an "Authorization Response" or "AuthRep Response" (right now it is both of them AuthRep and both of them Request).

Reply:
right, I meant to say as in the request is authorised `Authorize request` is the result

Reply:
hm? so the output is an Authorize _request_ rather than a response?

Reply:
no, the action at the end is to authorize the users request. I'm happy to change the terminology though if you find it confusing. In fact, if you find it confusing I'm sure others will so better I'll change it

Reply:
you could say "Authorize Response" or if it is asserting whether it is authorized or not authorized, "Authorize Response OK/NOK"